src/PgContractCache.ts

```diff
@@ -0,0 +1,218 @@
+import {
+  BasicSortKeyCache,
+  CacheKey,
+  CacheOptions,
+  LoggerFactory,
+  PruneStats,
+  SortKeyCacheResult,
+} from "warp-contracts";
+import { Client } from "pg";
+import { PgCacheOptions } from "./PgCacheOptions";
+
+export class PgContractCache<V> implements BasicSortKeyCache<V> {
+  private readonly logger = LoggerFactory.INST.create(PgContractCache.name);
+
+  private readonly client: Client;
+
+  constructor(
+    private readonly cacheOptions: CacheOptions,
+    private readonly pgCacheOptions?: PgCacheOptions
+  ) {
+    if (!pgCacheOptions) {
+      this.pgCacheOptions = {
+        maxEntriesPerContract: 10,
+      };
+    }
+    this.client = new Client(pgCacheOptions);
+  }
+
+  private async sortKeyTable() {
+    await this.client.query(
+      `
+          CREATE TABLE IF NOT EXISTS sort_key_cache
+          (
+              id       bigserial,
+              key      TEXT NOT NULL,
+              sort_key TEXT NOT NULL,
+              value    JSONB,
+              UNIQUE (key, sort_key)
+          );
+      `
+    );
+    await this.client.query(
+      "CREATE INDEX IF NOT EXISTS idx_sort_key_cache_key_sk ON sort_key_cache (key, sort_key);"
+    );
+  }
+
+  async begin(): Promise<void> {
+    await this.client.query("BEGIN;");
+  }
+
+  async close(): Promise<void> {
+    await this.client.end();
+    this.logger.info(`Connection closed`);
+    return;
+  }
+
+  async commit(): Promise<void> {
+    await this.client.query("COMMIT;");
+  }
+
+  async delete(key: string): Promise<void> {
+    await this.client.query("DELETE FROM sort_key_cache WHERE key = $1;", [
+      key,
+    ]);
+  }
+
+  dump(): Promise<any> {
+    return Promise.resolve(undefined);
+  }
+
+  async get(cacheKey: CacheKey): Promise<SortKeyCacheResult<V> | null> {
+    const result = await this.client.query(
+      `SELECT value
+       FROM sort_key_cache
+       WHERE key = $1
+         AND sort_key = $2;`,
+      [cacheKey.key, cacheKey.sortKey]
+    );
+
+    if (result && result.rows.length > 0) {
+      return new SortKeyCacheResult<V>(cacheKey.sortKey, result.rows[0].value);
+    }
+    return null;
+  }
+
+  async getLast(key: string): Promise<SortKeyCacheResult<V> | null> {
+    const result = await this.client.query(
+      "SELECT sort_key, value FROM sort_key_cache WHERE key = $1 ORDER BY sort_key DESC LIMIT 1",
```

Comment:
Add an index for this query

Reply:
Now there is an index for
 - `(key, sort_key DESC)`
 - `key`
 - `sort_key DESC`
 I think this is good enough?